Tech preview: Multinode, multi GPU deployment (#560)
opendatahub-io/opendatahub-documentation: modules/deploying-models-using-multiple-gpu-nodes.adoc

* You have enabled the operators for your GPU type, such as Node Feature Discovery Operator, NVIDIA GPU Operator. For more information about enabling accelerators, see link:{odhdocshome}/working-with-accelerators[Working with accelerators^].
endif::[]

** You are using an NVIDIA GPU (`nvidia.com/gpu`).
** Specify the GPU type through either the `ServingRuntime` or `InferenceService`. If the GPU type specified in the `ServingRuntime` differs from what is set in the `InferenceService`, both GPU types are assigned to the resource and can cause errors. 
* You have specified `serving.kserve.io/deploymentMode: RawDeployment` in the `InferenceService`. Your cluster does not have to be fully configured for KServe raw deployment.
* You have only 1 head pod in your setup. Do not adjust the replica count using the `min_replicas` or `max_replicas` settings in the `InferenceService`. Creating additional head pods can cause them to be excluded from the Ray cluster.	
* You have a Persistent Volume Claim (PVC) set up and configured for ReadWriteMany (RWX) access mode. 
* You have set the storage protocol for the `StorageURI` to `PVC`.
* Your Autoscaler is configured as `external`.

> NOTE: Ensure that `TENSOR_PARALLEL_SIZE` and `PIPELINE_PARALLEL_SIZE` are not already configured through environment variables.

Reviewer: this is not needed

.Procedure
. In a terminal window, if you are not already logged in to your {openshift-platform} cluster as a cluster administrator, log in to the {openshift-platform} CLI as shown in the following example:
+
[source]
----
$ oc login <openshift_cluster_url> -u <admin_username> -p <password>
----
+

. Select or create a namespace where you would like to deploy the model. In this example, `kserve-demo` is used and can be created using the following command:
+
[source]
----
oc new-project kserve-demo
----
+

> 

Reviewer: this is where pvc creation should be

+
[source]
----
... 
 predictor:
    model:
      runtime: vllm-multinode-runtime
      modelFormat:
        name: vLLM
      storageUri: pvc://llama-3-8b-pvc/hf/8b_instruction_tuned  
    workerSpec: {} # Specifying workerSpec indicates that multi-node functionality will be used    
----
+
After you create the `vllm-multinode-runtime.yaml` file, the following configurations are added to the serving runtime:
+
* `workerSpec.tensorParallelSize`: Determines how many GPUs are used per node. The GPU type count in both the head and worker node deployment resources is updated automatically. Ensure that the value of `workerSpec.tensorParallelSize` is at least `1`.
* `workerSpec.pipelineParallelSize`: Determines how many nodes are involved in the deployment. This variable represents the total number of nodes, including both the head and worker nodes. Ensure that the value of `workerSpec.pipelineParallelSize` is at least `2`.
+

. From the namespace where you would like to deploy the model, create a PVC for model storage and provide the name of your storage class. The storage class must be file storage.
+

> NOTE: If you have already configured a persistent volume claim (PVC), you can skip this step.

Reviewer: this step should be after creating a new project

+
[source]
----
kubectl apply -f - 
apiVersion: v1
kind: PersistentVolumeClaim
metadata:

>   name: llama-3-8b-pvc

Reviewer: ```suggestion
  name: granite-8b-code-base-pvc
```
to match the exmaple we provide

spec:
  accessModes:
    - ReadWriteMany
  volumeMode: Filesystem
  resources:
    requests:
      storage: 50Gi
  storageClassName: __<fileStorageClassName>__
----
+

. Download the model by exporting your token as an environment variable. In this example, `HF_TEST_TOKEN` is the exported token:
+
[source]
----
export HF_TEST_TOKEN=__<token>__
export MODEL=meta-llama/Meta-Llama-3-8B-Instruct

curl -o download-model-to-pvc.yaml https://raw.githubusercontent.com/kserve/website/a5a8661de3c3bd55b1faa1acf9125d05b10bfefd/docs/modelserving/v1beta1/llm/huggingface/multi-node/download-model-to-pvc.yaml

> envsubst < download-model-to-pvc.yaml | kubectl create -f -

Reviewer: should go after pvc (after project creation)

Download the model to the PVC, for example:
```
apiVersion: v1
kind: Pod
metadata:
  name: download-granite-8b-code
  labels:
    name: download-granite-8b-code
spec:
  volumes:
    - name: model-volume
      persistentVolumeClaim:
        claimName: granite-8b-code-claim
  restartPolicy: Never
  initContainers:
    - name: fix-volume-permissions
      image: quay.io/quay/busybox@sha256:92f3298bf80a1ba949140d77987f5de081f010337880cd771f7e7fc928f8c74d 
      command: ["sh"]
      args: ["-c", "mkdir -p /mnt/models/granite-8b-code-base && chmod -R 777 /mnt/models"]
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
  containers:
    - resources:
        requests:
          memory: 40Gi
      name: download-model
      imagePullPolicy: IfNotPresent
      image: quay.io/modh/kserve-storage-initializer@sha256:83ab292e87f2d30c5db4f756dac7d99d0a0252a4563f784f30a2d75cbbd46330
      args:
        - 's3://$(BUCKET_NAME)/granite-8b-code-base/'
        - /mnt/models/granite-8b-code-base
      env:
        - name: AWS_ACCESS_KEY_ID
          value: <id>
        - name: AWS_SECRET_ACCESS_KEY
          value: <secret>
        - name: BUCKET_NAME
          value: <bucket>
        - name: S3_USE_HTTPS
          value: "1"
        - name: AWS_ENDPOINT_URL
          value: <AWS endpoint>
        - name: awsAnonymousCredential
          value: 'false'
        - name: AWS_DEFAULT_REGION
          value: <region>
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
```

          value: <id>
        - name: AWS_SECRET_ACCESS_KEY
          value: <secret>
        - name: BUCKET_NAME
          value: <bucket_name>
        - name: S3_USE_HTTPS
          value: "1"
        - name: AWS_ENDPOINT_URL
          value: <AWS endpoint>
        - name: awsAnonymousCredential
          value: 'false'
        - name: AWS_DEFAULT_REGION
          value: <region>
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
----
+

. Create the `vllm-multinode-runtime` ServingRuntime as a custom runtime:
+
[source]
----
oc process vllm-multinode-runtime-template -n redhat-ods-applications|oc apply  -f -
----
+

. Deploy the model using the following `InferenceService` configuration:
+
[source]
----
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  annotations:
    serving.kserve.io/deploymentMode: RawDeployment
    serving.kserve.io/autoscalerClass: external
  name: granite-8b-code-base-pvc
spec:
  predictor:
    model:
      modelFormat:
        name: vLLM
      runtime: vllm-multinode-runtime
      storageUri: pvc://granite-8b-code-base-pvc/hf/8b_instruction_tuned
    workerSpec: {}    
----
+

> The following configuration can be added to the InferenceService:

Reviewer 2: Backticks for InferenceService?

+
* `workerSpec.tensorParallelSize`: Determines how many GPUs are used per node. The GPU type count in both the head and worker node deployment resources is updated automatically. Ensure that the value of `workerSpec.tensorParallelSize` is at least `1`.
* `workerSpec.pipelineParallelSize`: Determines how many nodes are involved in the deployment. This variable represents the total number of nodes, including both the head and worker nodes. Ensure that the value of `workerSpec.pipelineParallelSize` is at least `2`.

.Verification

> To confirm that you have set up your environment to deploy models on multiple GPU nodes, check the GPU resource status, the InferenceService status, the ray cluster status, and send a request to the model.

Reviewer 2: Backticks for InferenceService?

+-----------------------------------------+------------------------+----------------------+
         ...                                                               
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A10G                    On  |   00000000:00:1E.0 Off |                    0 |
|  0%   30C    P0             69W /  300W |18959MiB /  23028MiB <2>|      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+        
----
+

> Confirm that the model loaded properly by checking the values of <1> and <2>. If the model did not load, the value of these fields is `0MiB`.

Reviewer 2: Suggestion: Might want to validate how the callouts render upstream

> * Verify the status of your InferenceService using the following command:

Reviewer 2: Backticks for InferenceService

+

> NOTE: In the Technology Preview, you can only use port forwarding for inferencing.

Reviewer 2: Conditional for this note, since we mention that it is TP downstream?

|=========================================+========================+======================|
|   0  NVIDIA A10G                    On  |   00000000:00:1E.0 Off |                    0 |
|  0%   30C    P0             69W /  300W |18959MiB /  23028MiB <2>|      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+        
----
+
Confirm that the model loaded properly by checking the values of <1> and <2>. If the model did not load, the value of these fields is `0MiB`.

=== Check InferenceService status
. To verify the status of your InferenceService, run the following command:
+
[source]
----
oc wait --for=condition=ready pod/${podName} -n $DEMO_NAMESPACE --timeout=300s
export isvc_url=$(oc get route -n $DEMO_NAMESPACE |grep $MODEL_NAME| awk '{print $2}')
----
+

.Sample response
+
[source]
----
   NAME                 URL                                                   READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                          AGE
    vllm-llama3-8b   http://vllm-llama3-8b.default.example.com   
----

> +

Author: What note should be here? Is it related to this comment: we don't have rawDeployment mode feature for URL change so we will use port-forward.

If so, would something like this work?
> NOTE: In the Technology Preview, you can only use port-forwarding for inferencing.

> +

Reviewer 3: yes. it is the same issue. The `example.com` domain does look right for users so I want to add a note something like you just shared with the above comment

For more information about the support scope of {org-name} Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview[Technology Preview Features Support Scope]
====
endif::[]

.Prerequisites

* You have cluster administrator privileges for your {openshift-platform} cluster.
* You have downloaded and installed the {openshift-platform} command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].

ifndef::upstream[]
* You have enabled the operators for your GPU type, such as Node Feature Discovery Operator, NVIDIA GPU Operator, and so on. For more information about enabling accelerators, see link:{rhoaidocshome}{default-format-url}/managing_openshift_ai/working_with_accelerators[Working with accelerators^].
endif::[]
ifdef::upstream[]
* You have enabled the operators for your GPU type, such as Node Feature Discovery Operator, NVIDIA GPU Operator, and so on. For more information about enabling accelerators, see link:{odhdocshome}/working-with-accelerators[Working with accelerators^].
endif::[]

> ** You are using one of the following GPU types: `nvidia.com/gpu` (default), `intel.com/gpu`, `amd.com/gpu`, `habana.ai/gaudi`, or you have specified another GPU type in the annotations of the `InferenceService`. 

Reviewer: only `nvidia.com/gpu` was tested so I think we should remove the rest. @Jooho wdyt?

Reviewer 3: To me, it should work with various gpu type if they are using sole gpu type but I think we can add that the current version only tested with `nvidia.com/gpu`.

Author: I think I will remove other GPU types since only Nvidia was tested.

** Specify the GPU type through either the `ServingRuntime` or `InferenceService`. If the GPU type specified in the `ServingRuntime` differs from what is set in the `InferenceService`, both GPU types are assigned to the resource and can cause errors. 

ifndef::upstream[]

> * You have set up model-serving in raw deployment mode. For more information about raw deployment, see link:{rhoaidocshome}{default-format-url}serving_models/serving-large-models_serving-large-models#deploying-models-on-single-node-openshift-using-kserve-raw-deployment-mode_serving-large-models[Deploying models on single node {openshift-platform} using KServe raw deployment mode]

Reviewer: @Jooho I do not see this as a prerequisite; the isvc must have `serving.kserve.io/deploymentMode: RawDeployment` but it does not need to be cluster config.

Author: I believe the procedure linked here has the user update the isvc for raw deployment: https://docs.redhat.com/en/documentation/red_hat_openshift_ai_self-managed/2-latest/html/serving_models/serving-large-models_serving-large-models#deploying-models-on-single-node-openshift-using-kserve-raw-deployment-mode_serving-large-models
Let me know if the request here is for a different thing and I can make the update.

Reviewer 3: right. RawDeployment can be set by isvc. @eturner24 I think we can add that multi-node servingruntime is only supported in RawDeployment.

endif::[]
ifdef::upstream[]
* You have set up model-serving in raw deployment mode.
endif::[]

> * You have only 1 head pod in your setup. The replica count for the head pod can be adjusted using the `min_replicas` or `max_replicas` settings in the `InferenceService`. Creating additional head pods can cause them to be excluded from the Ray cluster.	

Reviewer: ```
The replica count for the head pod should not be adjusted using the `min_replicas` or `max_replicas` settings in the `InferenceService`
```
I think this should be the text; i.e not to use the replicas mechanism in the servingruntime.

* You have a Persistent Volume Claim (PVC) set up and configured for ReadWriteMany (RWX) access mode. 
* You have set the storage protocol for the `StorageURI` to `PVC`.
* Your Autoscaler is configured as `external`.

== Creating the `vllm-multinode-runtime` ServingRuntime template

Add the `vllm-multinode-runtime` ServingRuntime as a custom runtime to enable multi-node inferencing. The `vllm-multinode-runtime` ServingRuntime uses the same vLLM image as the `vllm-runtime` but it includes information necessary for multi-GPU inferencing.

> NOTE: Ensure that `TENSOR_PARALLEL_SIZE` and `PIPELINE_PARALLEL_SIZE` are not already configured through environment variables.

Reviewer: I would direct the user to use the provided ` vllm-multinode-runtime-template` template.
See following comment on how to use it.

.Procedure
. In a terminal window, if you are not already logged in to your {openshift-platform} cluster as a cluster administrator, log in to the {openshift-platform} CLI as shown in the following example:
+
[source]
----
$ oc login <openshift_cluster_url> -u <admin_username> -p <password>
----
+
. Select or create a namespace where you would like to deploy the model. In this example, `kserve-demo` is used and can be created using the following command:
+
[source]
----
oc new-project kserve-demo
----

> +

Reviewer: Create ServingRuntime using the multi node template:
```
oc process vllm-multinode-runtime-template -n redhat-ods-applications|oc apply  -f -
```

A `vllm-multinode-runtime` ServingRuntime will be created.

> . From the namespace, define a ServingRuntime object in a YAML file called `vllm-multinode-runtime.yaml` with the following contents:

Reviewer: ```suggestion
. From the namespace, define an InferenceService object in a YAML file called `vllm-multinode.yaml` with the following contents:
```

. Download the model by exporting your token as an environment variable. In this example, `HF_TEST_TOKEN` is the exported token:
+
[source]
----
export HF_TEST_TOKEN=__<token>__
export MODEL=meta-llama/Meta-Llama-3-8B-Instruct

curl -o download-model-to-pvc.yaml https://raw.githubusercontent.com/kserve/website/a5a8661de3c3bd55b1faa1acf9125d05b10bfefd/docs/modelserving/v1beta1/llm/huggingface/multi-node/download-model-to-pvc.yaml
envsubst < download-model-to-pvc.yaml | kubectl create -f -
----
+

ifndef::upstream[]
. From {productname-long}, create a custom model serving runtime in your data science project. For more information, see link:{rhoaidocshome}{default-format-url}serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime].
endif::[]
ifdef::upstream[]
. From {productname-long}, create a custom model serving runtime in your data science project. For more information, see link:{odhdocshome}/serving-models/#adding-a-custom-model-serving-runtime-for-the-multi-model-serving-platform_model-serving[Adding a custom model-serving runtime].
endif::[]

> . Modify the serving runtime with the path to your `vllm-multinode-runtime.yaml` file.

Reviewer: is this needed? the flow is to use a template

Author: @Jooho What do you think?

Reviewer 3: The template is not shown up in the dashboard so we would provide a yaml file. by the way, where is `vllm-multinode-runtime.yaml`file path? 

I think we can provide a way to create this template in user namespace for who can not access to redhat-ods-applications namespace. like the following:

```
curl -L https://raw.githubusercontent.com/opendatahub-io/odh-model-controller/refs/heads/main/config/runtimes/vllm-multinode-template.yaml -o /tmp/vllm-multinode-template.yaml

cat /tmp/vllm-multinode-template.yaml |sed "s+image: .*$+image: quay.io/modh/vllm@sha256:ce5a8e5fc91442229ea4f54ea251cf65f10d383021138f8f0d55dcc5c3eb038b+g"  |oc apply -f -
```

Reviewer: the template is available on the cluster and it was decided to let only admins use it so i think we should go with `oc process`
i also think we should not point to the template in the repo as it may be updated but not tested yet in d/s.

Author: Just so I'm clear, should we leave this step as-is or are there changes I should make here?

Reviewer 3: @rnetser This is the one that tarun brought up last time. have you tested to create the servingruntime through template with normal user?

Reviewer: A normal user cannot access the template and we agreed that only an admin can do it.

      imagePullPolicy: IfNotPresent
      image: quay.io/modh/kserve-storage-initializer@sha256:83ab292e87f2d30c5db4f756dac7d99d0a0252a4563f784f30a2d75cbbd46330
      args:
        - 's3://$(BUCKET_NAME)/granite-8b-code-base/'
        - /mnt/models/granite-8b-code-base
      env:
        - name: AWS_ACCESS_KEY_ID
          value: <id>
        - name: AWS_SECRET_ACCESS_KEY
          value: <secret>
        - name: BUCKET_NAME
          value: <bucket>
        - name: S3_USE_HTTPS
          value: "1"
        - name: AWS_ENDPOINT_URL
          value: <AWS endpoint>
        - name: awsAnonymousCredential
          value: 'false'
        - name: AWS_DEFAULT_REGION
          value: <region>
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
----
+

. Create the `vllm-multinode-runtime` ServingRuntime template as follows:

.. Add the `vllm-multinode-runtime` ServingRuntime as a custom runtime:
+
[source]
----
oc process vllm-multinode-runtime-template -n redhat-ods-applications|oc apply  -f -
----
+

.. From the namespace, define an InferenceService object in a YAML file called `vllm-multinode.yaml` with the following contents:
+
[source]
----
... 

>  predictor:

Reviewer 3: same : https://github.com/opendatahub-io/opendatahub-documentation/pull/560#discussion_r1869623144

    model:
      runtime: vllm-multinode-runtime
      modelFormat:
        name: vLLM

>       storageUri: pvc://granite-8b-code-base-pvc/hf/8b_instruction_tuned  

Reviewer 3: I think we already discussed it in the slack. @rnetser do we have a public repository that can provide this model? 
If not, we can not use this model for doc.

    workerSpec: {} # Specifying workerSpec indicates that multi-node functionality will be used    
----
+

> After you create the `vllm-multinode-runtime.yaml` file, the following configurations are added to the serving runtime:

Reviewer 3: This part looks 2 things mingled. 
1. Creating `vllm-multinode-runtime.yaml` that is ServingRuntime
2. `oc process ..` . this is creating ServingRuntime with template.

Anyway, this line would explain that user can edit the following a new API for mult-node in the ServingRuntime as a  default value. But it can be overwritten by InferenceService.

+
* `workerSpec.tensorParallelSize`: Determines how many GPUs are used per node. The GPU type count in both the head and worker node deployment resources is updated automatically. Ensure that the value of `workerSpec.tensorParallelSize` is at least `1`.
* `workerSpec.pipelineParallelSize`: Determines how many nodes are involved in the deployment. This variable represents the total number of nodes, including both the head and worker nodes. Ensure that the value of `workerSpec.pipelineParallelSize` is at least `2`.
+

> ifndef::upstream[]

Reviewer 3: This is explanation for a new API how to set nodes counts and what is minimum value for it.

> ifndef::upstream[]

Reviewer: the comment is on line 169 -onwards (without the verification section)

      image: quay.io/quay/busybox@sha256:92f3298bf80a1ba949140d77987f5de081f010337880cd771f7e7fc928f8c74d 
      command: ["sh"]
      args: ["-c", "mkdir -p /mnt/models/granite-8b-code-base && chmod -R 777 /mnt/models"]
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
  containers:
    - resources:
        requests:
          memory: 40Gi
      name: download-model
      imagePullPolicy: IfNotPresent
      image: quay.io/modh/kserve-storage-initializer@sha256:83ab292e87f2d30c5db4f756dac7d99d0a0252a4563f784f30a2d75cbbd46330
      args:
        - 's3://$(BUCKET_NAME)/granite-8b-code-base/'
        - /mnt/models/granite-8b-code-base
      env:
        - name: AWS_ACCESS_KEY_ID
          value: <id>
        - name: AWS_SECRET_ACCESS_KEY
          value: <secret>
        - name: BUCKET_NAME
          value: <bucket>
        - name: S3_USE_HTTPS
          value: "1"
        - name: AWS_ENDPOINT_URL
          value: <AWS endpoint>
        - name: awsAnonymousCredential
          value: 'false'
        - name: AWS_DEFAULT_REGION
          value: <region>
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
----
+

> . Create the `vllm-multinode-runtime` ServingRuntime template as follows:

Reviewer 3: ```suggestion

```

> .. Add the `vllm-multinode-runtime` ServingRuntime as a custom runtime:

Reviewer 3: ```suggestion
. Create the `vllm-multinode-runtime` ServingRuntime as a custom runtime:
```

+
[source]
----
oc process vllm-multinode-runtime-template -n redhat-ods-applications|oc apply  -f -
----
+

.. From the namespace, define an InferenceService object in a YAML file called `vllm-multinode.yaml` with the following contents:
+
[source]
----

> apiVersion: serving.kserve.io/v1beta1

Reviewer: the indentation is wrong

```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: vllm-multi
  annotations:
    serving.kserve.io/deploymentMode: RawDeployment
    serving.kserve.io/autoscalerClass: external  
spec:
  predictor:
    model:
      modelFormat:
        name: vLLM
      runtime: vllm-multinode-runtime
      storageUri: pvc://granite-8b-code-claim/granite-8b-code-base
    workerSpec: {}
```

* You have downloaded and installed the {openshift-platform} command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].

ifndef::upstream[]
* You have enabled the operators for your GPU type, such as Node Feature Discovery Operator, NVIDIA GPU Operator. For more information about enabling accelerators, see link:{rhoaidocshome}{default-format-url}/managing_openshift_ai/working_with_accelerators[Working with accelerators^].
endif::[]
ifdef::upstream[]
* You have enabled the operators for your GPU type, such as Node Feature Discovery Operator, NVIDIA GPU Operator. For more information about enabling accelerators, see link:{odhdocshome}/working-with-accelerators[Working with accelerators^].
endif::[]

** You are using an NVIDIA GPU (`nvidia.com/gpu`).
** Specify the GPU type through either the `ServingRuntime` or `InferenceService`. If the GPU type specified in the `ServingRuntime` differs from what is set in the `InferenceService`, both GPU types are assigned to the resource and can cause errors. 

> * You have specified `serving.kserve.io/deploymentMode: RawDeployment` in the `InferenceService`. Your cluster does not have to be fully configured for KServe raw deployment.

Author: @rnetser @Jooho 
Does this update work here?

Reviewer: maybe rephrase to `make sure your cluster is configured with kserve enabled`?
